File: app/config/host_snapshot.py
"""进程级主机环境快照：启动时采集一次，供后续逻辑直接读取。"""

from __future__ import annotations

import getpass
import logging
import os
import platform
import sys
import time
from dataclasses import dataclass

_logger = logging.getLogger(__name__)

_snapshot: HostSnapshot | None = None


@dataclass(frozen=True, slots=True)
class HostSnapshot:
    """启动时刻采集的环境信息（不可变）。

    字段说明：
    - **os_kind**：粗分类，`windows` / `linux` / `darwin` / `other`（与 `sys.platform` 推断一致）；
    - **login_name**：`getpass.getuser()`，失败则为空串；
    - **effective_uid / effective_gid**：POSIX 有效 UID/GID；Windows 等非 POSIX 为 `None`。
    """

    captured_at_unix: float
    os_kind: str
    sys_platform: str
    platform_system: str
    platform_release: str
    machine: str
    login_name: str
    effective_uid: int | None
    effective_gid: int | None


def _infer_os_kind() -> str:
    """与 `detect_host_os` 对齐的粗分类字符串（避免 config 依赖 harness）。"""

    p = sys.platform
    if p == "win32" or p.startswith("cygwin") or p == "msys":
        return "windows"
    if p == "darwin":
        return "darwin"
    if p.startswith("linux"):
        return "linux"
    return "other"


def _build_host_snapshot() -> HostSnapshot:
    """构造当前进程快照（单次探针）。"""

    try:
        login_name = str(getpass.getuser() or "").strip()
    except Exception:
        login_name = ""
    euid: int | None = None
    egid: int | None = None
    try:
        euid = int(os.geteuid())
        egid = int(os.getgid())
    except (AttributeError, OSError, TypeError, ValueError):
        euid = None
        egid = None
    return HostSnapshot(
        captured_at_unix=time.time(),
        os_kind=_infer_os_kind(),
        sys_platform=sys.platform,
        platform_system=platform.system(),
        platform_release=platform.release(),
        machine=platform.machine(),
        login_name=login_name,
        effective_uid=euid,
        effective_gid=egid,
    )


def capture_host_snapshot_at_startup() -> HostSnapshot:
    """在 API 启动路径显式采集并缓存快照，并打一条 INFO 日志。

    逻辑：
    1. 调用 **`_build_host_snapshot`** 写入模块级缓存；
    2. 输出结构化 **`logging`** 记录，便于运维检索；
    3. 返回同一快照对象。

    关键分支：
    - 可重复调用：每次覆盖缓存（适用于调试重启；正式环境通常只调一次）。

    副作用：
    - 修改模块级 **`_snapshot`**。
    """

    global _snapshot
    _snapshot = _build_host_snapshot()
    snap = _snapshot
    _logger.info(
        "[dagents][host_snapshot] os_kind=%s login_name=%r euid=%s egid=%s sys_platform=%r "
        "platform_system=%r platform_release=%r machine=%r captured_at_unix=%s",
        snap.os_kind,
        snap.login_name,
        snap.effective_uid,
        snap.effective_gid,
        snap.sys_platform,
        snap.platform_system,
        snap.platform_release,
        snap.machine,
        snap.captured_at_unix,
    )
    return snap


def get_host_snapshot() -> HostSnapshot:
    """返回主机快照；若尚未显式采集则惰性构建并缓存（不写 INFO 启动日志）。

    逻辑：
    1. 缓存命中直接返回；
    2. 否则 **`_build_host_snapshot`** 写入缓存再返回（单测、工具先于 API 调用时用）。

    副作用：
    - 可能在首次调用时写入 **`_snapshot`**。
    """

    global _snapshot
    if _snapshot is None:
        _snapshot = _build_host_snapshot()
    return _snapshot

File: node/internal/hostsnapshot/hostsnapshot.go
// Package hostsnapshot 提供进程级主机环境快照（启动时采集，供 system prompt 等读取）。
package hostsnapshot

import (
	"fmt"
	"os"
	"os/user"
	"runtime"
	"strings"
	"sync"
	"time"
)

// Snapshot 为启动时刻采集的环境信息（不可变）。
type Snapshot struct {
	CapturedAtUnix  float64
	OSKind          string
	SysPlatform     string
	PlatformSystem  string
	PlatformRelease string
	Machine         string
	LoginName       string
	EffectiveUID    *int
	EffectiveGID    *int
}

var (
	mu     sync.RWMutex
	cached *Snapshot
)

// CaptureAtStartup 显式采集并缓存快照；Node 启动路径调用一次。
func CaptureAtStartup() Snapshot {
	s := buildSnapshot()
	mu.Lock()
	cached = &s
	mu.Unlock()
	return s
}

// Get 返回主机快照；尚未采集时惰性构建。
func Get() Snapshot {
	mu.RLock()
	if cached != nil {
		s := *cached
		mu.RUnlock()
		return s
	}
	mu.RUnlock()
	s := buildSnapshot()
	mu.Lock()
	if cached == nil {
		cached = &s
	} else {
		s = *cached
	}
	mu.Unlock()
	return s
}

func buildSnapshot() Snapshot {
	loginName := ""
	if u, err := user.Current(); err == nil {
		loginName = strings.TrimSpace(u.Username)
	}
	var euid, egid *int
	if runtime.GOOS != "windows" {
		if uid := os.Getuid(); uid >= 0 {
			v := uid
			euid = &v
		}
		if gid := os.Getgid(); gid >= 0 {
			v := gid
			egid = &v
		}
	}
	return Snapshot{
		CapturedAtUnix:  float64(time.Now().Unix()),
		OSKind:          inferOSKind(),
		SysPlatform:     runtime.GOOS,
		PlatformSystem:  runtime.GOOS,
		PlatformRelease: runtime.Version(),
		Machine:         runtime.GOARCH,
		LoginName:       loginName,
		EffectiveUID:    euid,
		EffectiveGID:    egid,
	}
}

func inferOSKind() string {
	switch runtime.GOOS {
	case "windows":
		return "windows"
	case "darwin":
		return "darwin"
	case "linux":
		return "linux"
	default:
		return "other"
	}
}

// FormatEnvironmentSection 格式化为 system prompt「当前运行环境」正文。
func FormatEnvironmentSection(s Snapshot) string {
	loginDisplay := strings.TrimSpace(s.LoginName)
	if loginDisplay == "" {
		loginDisplay = "未知"
	}
	platformLine := "`" + s.SysPlatform + "` · " + s.PlatformSystem + " " + s.PlatformRelease + " · " + s.Machine
	lines := []string{
		"- 操作系统类别：`" + s.OSKind + "`",
		"- 平台摘要：" + platformLine,
		"- 当前进程用户（登录名）：`" + loginDisplay + "`",
	}
	if s.EffectiveUID != nil && s.EffectiveGID != nil {
		lines = append(lines, fmt.Sprintf("- 有效 UID / GID：`%d` / `%d`", *s.EffectiveUID, *s.EffectiveGID))
	} else {
		lines = append(lines, "- 有效 UID / GID：不适用（当前运行时非 POSIX 或未提供）")
	}
	return strings.Join(lines, "\n")
}
